### ISSUE TITLE
Fail to Transfer 10 (More than one) Pre minted tokens to multiple receivers in one Transaction (Airdrop Scenario)

### ISSUE THREAD

**Issue by edridudi** [bloxbean/cardano-client-lib#53]
(Using preview2 tag) 

below Code based on https://github.com/bloxbean/cardano-client-lib/blob/3d85b52d4984af369457fb2628c3a507e23a40fa/src/integration-test/java/com/bloxbean/cardano/client/backend/api/helper/TransactionHelperServiceIT.java#L267

Returned with the error:

`Result{successful=false, response='{"error":"Bad Request","message":"\"transaction read error RawCborDecodeError [DecoderErrorDeserialiseFailure \\\"Byron Tx\\\" (DeserialiseFailure 1 \\\"Size mismatch when decoding TxAux.\\\\nExpected 2, but found 4.\\\"),DecoderErrorDeserialiseFailure \\\"Shelley Tx\\\" (DeserialiseFailure 1 \\\"Size mismatch when decoding \\\\nRecord RecD.\\\\nExpected 4, but found 3.\\\"),DecoderErrorDeserialiseFailure \\\"Shelley Tx\\\" (DeserialiseFailure 1 \\\"Size mismatch when decoding \\\\nRecord RecD.\\\\nExpected 4, but found 3.\\\"),DecoderErrorDeserialiseFailure \\\"Shelley Tx\\\" (DeserialiseFailure 1 \\\"Size mismatch when decoding \\\\nRecord RecD.\\\\nExpected 4, but found 3.\\\"),DecoderErrorDeserialiseFailure \\\"Shelley Tx\\\" (DeserialiseFailure 1207 \\\"expected word\\\")]\"","status_code":400}', code=400, value=null}`

```
public boolean transferMultiAssets() throws ApiException, AddressExcepion, CborSerializationException {
        List<PaymentTransaction> paymentTransactionList = new ArrayList<>();
        List<String> receivers = List.of("addr_test1q825tw2qpyfvpnmdxrc4ythrl3zq8dpk8ggwxxd230xndquvjw44r906qaszsqktacayq37zlavvft875g8afrd6mazsty96ya","addr_test1q8348ly466yurmutrkg4whmgvdkwnqczgr2zq3xgtv42hhremeq9kwvjnkrmwpdl7xrc5pvdeffnzr42prwfjatvz6sq40jdwu","addr_test1q83agdaplctqkrjdke07a0mv98ptnu48s2w8mt44qklnggtheegtaefdf43548rlygf9d8y8lu380ggdkp2jmzhl7sas6l8lua","addr_test1q84rz3unwmytncxdqwmye8xryz9532e3uwwlajym3u2myngm7lvt636g2mzk4mq4m6jggap6yxwwdy7eysg94cnx527qfzsj0h","addr_test1q856rzrmugaqxvpkft7zlt70kn47m99m7h0u45dxdu3tv8k80q0afh83fjl7mvuapx54h0r4xy5tch0anym7c5m953yst3g7uy","addr_test1q85ytct28gef68cja6vlmyhachdeec2h5zvtmzq0gj2cd0gk02mux9uvtkam3kg90uw74fdsr7tf2wnwr7y8e4hz2kwskm7k0w","addr_test1q87a345pfhgnwlh3czvr3y6zs6cad3vhcgact8jg06s0ufyn8ua6wq9tl9rgrq6ae7mvwf5jnk952jkylkaennw26fnq6qk952","addr_test1q87q68uzn6u9a8htm5clxekygsd03hgfdst0su5phyul7gc9zuyzrg464vzrgzazrfz606h8qgx5ay3usnfc9m76wgssd680hh","addr_test1q88dlzs7pu6jc6sfh072v3lzpeehcwxse07ftlr0vu9kdyfa38wnqjekyhvhrjmw25skcydqa4plx0z6c3lasawzcakqx05nnr");
        Random rand = new Random();
        for (String address : receivers) {
            String unit;
            if (rand.nextInt(100)<50) {
                unit = "9668ef339ea4b29a29b7a500b1a1f6749568ddb623cc463f95fe07f24d75736963476c6f62654f6e65";
            } else {
                unit = "9668ef339ea4b29a29b7a500b1a1f6769568ddb6232c463f95fe07f24d75736963476c6f626554776f";
            }
            PaymentTransaction paymentTransaction = PaymentTransaction.builder().sender(walletService.getAccountByWalletIndex(0)).receiver(address).amount(BigInteger.valueOf(1)).unit(unit).build();
            paymentTransactionList.add(paymentTransaction);
        }
        //Calculate total fee for all 3 payment transactions and set in one of the payment transaction
        BigInteger fee = cardanoApiService.getFeeCalculationService().calculateFee(paymentTransactionList, TransactionDetailsParams.builder().ttl(cardanoApiService.getTtl()).build(), null);
        paymentTransactionList.get(0).setFee(fee);
        Result<TransactionResult> result = cardanoApiService.getTransactionHelperService().transfer(paymentTransactionList, TransactionDetailsParams.builder().ttl(cardanoApiService.getTtl()).build());
        log.info(result.toString());
        return result.isSuccessful();
}
```

**Comment by satran004** [comment]
@edridudi Just gave a quick try with 4 different tokens including Ada. It seems to be working for me.  

I will try with 10 tokens later. 

Meanwhile, can you please enable DEBUG log level and see if you can capture the final transaction JSON in log to find the root cause of de-serialization error ?

Link to my test:

Txn hash: https://testnet.cardanoscan.io/transaction/225420de31f3689605b37a4ae8697a104bda5f39c0bf4e5d88541c1c982608aa

Test code: https://gist.github.com/satran004/56f54b553e49d4085583aec318eb45fe

**Comment by igurv** [comment]
> @edridudi Just gave a quick try with 4 different tokens including Ada. It seems to be working for me.
> 
> I will try with 10 tokens later.
> 
> Meanwhile, can you please enable DEBUG log level and see if you can capture the final transaction JSON in log to find the root cause of de-serialization error ?
> 
> Link to my test:
> 
> Txn hash: https://testnet.cardanoscan.io/transaction/225420de31f3689605b37a4ae8697a104bda5f39c0bf4e5d88541c1c982608aa
> 
> Test code: https://gist.github.com/satran004/56f54b553e49d4085583aec318eb45fe

Hey there! I'm also not being successful to create an array of multiple transactions and submit it to the network.
Your code works flawlessly, I tested it, but when I try to create multiple transactions from a file, I get CBOR errors.

Can you please help me out? The code runs under you coding structure I can't figure out why it's not working.

Here's my code bellow:

```
package retro;

import com.bloxbean.cardano.client.account.Account;
import com.bloxbean.cardano.client.backend.api.*;
import com.bloxbean.cardano.client.backend.api.helper.FeeCalculationService;
import com.bloxbean.cardano.client.backend.api.helper.TransactionHelperService;
import com.bloxbean.cardano.client.backend.api.helper.model.TransactionResult;
import com.bloxbean.cardano.client.backend.exception.ApiException;
import com.bloxbean.cardano.client.backend.factory.BackendFactory;
import com.bloxbean.cardano.client.backend.impl.blockfrost.common.Constants;
import com.bloxbean.cardano.client.backend.model.Block;
import com.bloxbean.cardano.client.backend.model.Result;
import com.bloxbean.cardano.client.backend.model.TransactionContent;
import com.bloxbean.cardano.client.common.model.Networks;
import com.bloxbean.cardano.client.exception.AddressExcepion;
import com.bloxbean.cardano.client.exception.CborSerializationException;
import com.bloxbean.cardano.client.transaction.model.PaymentTransaction;
import com.bloxbean.cardano.client.transaction.model.TransactionDetailsParams;
import com.bloxbean.cardano.client.util.JsonUtil;

import java.io.File;
import java.io.IOException;
import java.math.BigInteger;
import java.util.Arrays;
import java.util.Scanner;


public class Droptest {

    private final TransactionService transactionService;
    private final UtxoService utxoService;
    private final TransactionHelperService transactionHelperService;
    private final BlockService blockService;
    private final FeeCalculationService feeCalculationService;
    private final EpochService epochService;

    public Droptest() {
        BackendService backendService = BackendFactory.getBlockfrostBackendService(Constants.BLOCKFROST_TESTNET_URL, "testnet7al9w7cUyHEYdk3VSpGEtenCNc88n3OH");
        utxoService = backendService.getUtxoService();
        transactionService = backendService.getTransactionService();
        transactionHelperService = backendService.getTransactionHelperService();
        blockService = backendService.getBlockService();
        feeCalculationService = backendService.getFeeCalculationService();
        epochService = backendService.getEpochService();
    }

    public static void main(String[] args) throws ApiException, AddressExcepion, CborSerializationException, IOException {

        Droptest droptest = new Droptest();

        String senderMnemonic = "1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24";
        Account sender = new Account(Networks.testnet(), senderMnemonic);
        PaymentTransaction[] translist = new PaymentTransaction[40];


        Scanner scanner = new Scanner(new File("file.txt"));

        for (int i = 0; i < 40; i++) {

            if (scanner.hasNextLine()) {
                String line = scanner.nextLine();
                String[] parts = line.split(",");

                if(i == 0) {
                    PaymentTransaction paymentTransaction =
                            PaymentTransaction.builder()
                                    .sender(sender)
                                    .receiver(parts[1])
                                    .unit(parts[0])
                                    .amount(BigInteger.valueOf(1))
                                    .fee(BigInteger.valueOf(4000))//some low fee (invalid). Just for testing
                                    .build();

                    translist[i] = paymentTransaction;

                } else{

                    PaymentTransaction paymentTransaction =
                            PaymentTransaction.builder()
                                    .sender(sender)
                                    .receiver(parts[1])
                                    .unit(parts[0])
                                    .amount(BigInteger.valueOf(1))
                                    .build();

                    translist[i] = paymentTransaction;

                }

            } else {
                System.out.println("Break dance!");
                break;
            }
        }

        BigInteger fee = droptest.feeCalculationService.calculateFee(Arrays.asList(translist),
                TransactionDetailsParams
                        .builder()
                        .ttl(droptest.getTtl())
                        .build(), null);
        translist[0].setFee(fee);

        System.out.println(Arrays.asList(translist));

        Result<TransactionResult> result = droptest.transactionHelperService.transfer(Arrays.asList(translist),
                TransactionDetailsParams.builder().ttl(droptest.getTtl()).build());

        if(result.isSuccessful())
            System.out.println("Transaction Id: " + result.getValue());
        else
            System.out.println("Transaction failed: " + result);

        System.out.println(result);
        droptest.waitForTransaction(result);
    }

    public void waitForTransaction(Result<TransactionResult> result) {
        try {
            if (result.isSuccessful()) { //Wait for transaction to be mined
                int count = 0;
                while (count < 60) {
                    Result<TransactionContent> txnResult = transactionService.getTransaction(result.getValue().getTransactionId());
                    if (txnResult.isSuccessful()) {
                        System.out.println(JsonUtil.getPrettyJson(txnResult.getValue()));
                        break;
                    } else {
                        System.out.println("Waiting for transaction to be mined ....");
                    }

                    count++;
                    Thread.sleep(2000);
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public long getTtl() throws ApiException {

        Block block = blockService.getLastestBlock().getValue();
        long slot = block.getSlot();
        return slot + 2000;
    }


}

```

**Comment by satran004** [comment]
@igurv Can you please post the cbor error here ?

**Comment by igurv** [comment]
> @igurv Can you please post the cbor error here ?

Hey there @satran004 I've tried the code above and it works now, with a problem.

Looks like max output are 4 tx.

This means that with the @edridudi code, I'm only able to send 1 ADA tx + 3 assets in the same transaction.

The error I get it:

```
Exception in thread "main" java.lang.NegativeArraySizeException: -3
	at com.bloxbean.cardano.client.util.AssetUtil.getPolicyIdAndAssetName(AssetUtil.java:29)
	at com.bloxbean.cardano.client.backend.api.helper.impl.UtxoTransactionBodyBuilder.toAsset(UtxoTransactionBodyBuilder.java:204)
	at com.bloxbean.cardano.client.backend.api.helper.impl.UtxoTransactionBodyBuilder.groupRequestedOutput(UtxoTransactionBodyBuilder.java:380)
	at com.bloxbean.cardano.client.backend.api.helper.impl.UtxoTransactionBodyBuilder.lambda$calculateOutputs$25(UtxoTransactionBodyBuilder.java:437)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.HashMap$EntrySpliterator.forEachRemaining(HashMap.java:1764)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578)
	at com.bloxbean.cardano.client.backend.api.helper.impl.UtxoTransactionBodyBuilder.calculateOutputs(UtxoTransactionBodyBuilder.java:438)
	at com.bloxbean.cardano.client.backend.api.helper.impl.UtxoTransactionBodyBuilder.buildTransferBody(UtxoTransactionBodyBuilder.java:98)
	at com.bloxbean.cardano.client.backend.api.helper.impl.UtxoTransactionBuilderImpl.buildTransaction(UtxoTransactionBuilderImpl.java:92)
	at com.bloxbean.cardano.client.backend.api.helper.TransactionHelperService.createSignedTransaction(TransactionHelperService.java:191)
	at com.bloxbean.cardano.client.backend.api.helper.impl.FeeCalculationServiceImpl.calculateFee(FeeCalculationServiceImpl.java:162)
	at com.bloxbean.cardano.client.backend.api.helper.impl.FeeCalculationServiceImpl.calculateFee(FeeCalculationServiceImpl.java:142)
	at com.bloxbean.cardano.client.Droptest.transferMultiAssetMultiPayments_whenSingleSender_multipleToken(Droptest.java:85)
	at com.bloxbean.cardano.client.Droptest.main(Droptest.java:50)
```
Please let me know if you need anything else from my side.

Best regards and thank you very much for helping out.

**Comment by satran004** [comment]
Thanks @igurv. I will check.

**Comment by satran004** [comment]
@igurv I am looking into this issue. I am now able to reproduce it. Btw, which lib version are you using ? 0.2.0-beta2 ?

**Comment by satran004** [comment]
@igurv 

I found few issues currently in the high-level api (PaymentTransaction) while sending more tokens from single sender. Somehow, there is a token in the change output with -1 value. This causes cbor error in my case.

I tried the same use case with new composable functions approach and it works as expected. But you need to use **0.2.0-beta2** version for  now.  We will try to fix this issue (For high level api) in the next release.

For now, you can refer to his gist for sample code using composable functions https://gist.github.com/satran004/e7014135d475446f5af2aa3da168717a

-  MintMultiple.java  : Mint 10 new NFT and add those to nfts.txt file with 10 random receivers
- DropTestFuctional :  Read nfts.txt file and drop tokens to the receivers


Here's a sample transaction with the same use case: https://testnet.cardanoscan.io/transaction/76642620e8c6fda0ba5a6050c9f839335c695d4ad349c8fda95418369ca279bc

Let me know if you face any issue.

**Comment by satran004** [comment]
@igurv 
BTW, I tried the same with high-level api with the current **SNAPSHOT** version (0.2.0-beta3-SNAPSHOT). It is working. So the bug seems to be fixed.

The 0.2.0-beta3 should be out in few days. 

There are some package changes and few minor api breaking changes in beta3. So if you  want to try with the SNAPSHOT version with high-level api, please check this gist for the code and dependency.

https://gist.github.com/satran004/6afbec7cb31998750a349951e888e6e9

**Comment by edridudi** [comment]
@satran004 Thank you very much.
The issue is resolved in 0.2.0-beta3-SNAPSHOT.
